docs/debugging.md

```diff
@@ -0,0 +1,741 @@
+# 1. Introduction
+
+Often when writing eBPF programs, the first attempt will fail verification.
+This tutorial illustrates how to understand and debug eBPF verification failures.
+
+If you're new to eBPF for Windows, we recommend first going through the [basic eBPF tutorial](tutorial.md).
+Once you understand that tutorial and have llvm-objdump and the netsh helper installed
+on your machine, you're ready for the debugging tutorial.
+
+# 2. Debugging a buggy eBPF program
+
+Let's start with the [droppacket_unsafe.c](../tests/sample/droppacket_unsafe.c) program, which
+is compiled as part of building eBPF for Windows, as it is used in the unit tests.
+
+**Step 1)** Let's first look at what sections are in the file:
+
+```
+> cd x64\Debug
+> netsh ebpf show sections droppacket_unsafe.o
+
+             Section       Type  # Maps    Size
+====================  =========  ======  ======
+               .text        xdp       1       3
+                 xdp        xdp       1      20
+```
+
+**Step 2)** Let's try to verify each section:
+
+```
+> netsh ebpf show ver droppacket_unsafe.o .text
+Verification failed
+
+Verification report:
+
+1: r0.type == number
+2: r0.type == number
+
+2 errors
+
+> netsh ebpf show ver droppacket_unsafe.o xdp
+Verification failed
+
+Verification report:
+
+2: Upper bound must be at most packet_size (valid_access(r1.offset+9, width=1))
+4: Upper bound must be at most packet_size (valid_access(r1.offset+24, width=2))
+
+2 errors
+
+```
+
+We can see that both sections have issues.   We'll look at these one at a time.
+
+**Step 3)** Let's first look at the netsh disassembly output for the .text section:
+
+```
+> netsh ebpf show disassembly droppacket_unsafe.o .text
+       0:       r0 = r1
+       1:       r0 = be16 r0
+       2:       exit
+
+```
+
+We see there are 3 instructions, numbered 0 through 2.  In eBPF programs,
+"r0" through "r10" are registers, where r10 is the stack pointer and
+r0 is used for return values. The hook context (ctx) structure
+pointer would be in r6, and at the start of the function is also in r1 since it is
```

Comment:
I am not convinced about the reference to r6 here. I believe this is an artifact of the C compiler you are using and not part of the BPF ABI.

R6->R9 are general purpose non-volatile registers that are preserved across calls to helper functions. The compiler could preserve the context pointer in any of R6->R9 or push it onto the stack. It just happens that it's often stored in R6 as that's the first non-volatile register.

Reply:
Updated, please re-review this paragraph